v1/format/json_marshaller.cc

```diff
@@ -0,0 +1,129 @@
+#include "json_marshaller.h"
+
+namespace cloud_events {
+namespace format {
+
+using ::io::cloudevents::v1::CloudEvent;
+using ::io::cloudevents::v1::CloudEvent_CloudEventAttribute;
+
+absl::StatusOr<Json::Value> JsonMarshaller::PrintToJson(CloudEvent_CloudEventAttribute attr){
+    switch (attr.attr_oneof_case()) {
+        case CloudEvent_CloudEventAttribute::AttrOneofCase::kCeBoolean:
+            return Json::Value(attr.ce_boolean());
+        case CloudEvent_CloudEventAttribute::AttrOneofCase::kCeInteger:
+            return Json::Value(attr.ce_integer());
+        case CloudEvent_CloudEventAttribute::AttrOneofCase::kCeString:
+            return Json::Value(attr.ce_string());
+        case CloudEvent_CloudEventAttribute::AttrOneofCase::kCeBinary:
+            return Json::Value(attr.ce_binary());
+        case CloudEvent_CloudEventAttribute::AttrOneofCase::kCeUri:
+            return Json::Value(attr.ce_uri());
+        case CloudEvent_CloudEventAttribute::AttrOneofCase::kCeUriReference:
+            return Json::Value(attr.ce_uri_reference());
+        case CloudEvent_CloudEventAttribute::AttrOneofCase::kCeTimestamp:
+            return Json::Value(google::protobuf::util::TimeUtil::ToString(attr.ce_timestamp()));
+        case CloudEvent_CloudEventAttribute::AttrOneofCase::ATTR_ONEOF_NOT_SET:
+            return absl::InvalidArgumentError("Cloud Event metadata attribute not set.");
+    }
+    return absl::InvalidArgumentError("Cloud Event metadata attribute not recognized"); // should not be reached as all enums covered
+}
+
+absl::StatusOr<StructuredCloudEvent> JsonMarshaller::Serialize(CloudEvent cloud_event) {
+    Json::Value root;
+
+    // handle metadata
+    if (cloud_event.id().empty() || cloud_event.source().empty() || cloud_event.spec_version().empty(), cloud_event.type().empty()) {
+        return absl::InvalidArgumentError("Required attribute in Cloud Event cannot be empty");
+    }
+
+    root["id"] = cloud_event.id();
+    root["source"] = cloud_event.source();
+    root["spec_version"] = cloud_event.spec_version();
+    root["type"] = cloud_event.type();
+
+    for (auto const& x : cloud_event.attributes()) {
+        absl::StatusOr<Json::Value> json_printed = JsonMarshaller::PrintToJson(x.second);
+        if (!json_printed) {
+            return json_printed.status();
+        } else {
+            root[x.first] = json_printed.value();
```

Comment:
prefer accessing values of StatusOr variables through either "" or "->" operators ("Accessing the value held by a StatusOr should be performed via operator or operator->, after a call to ok() has verified that the value is present.")

root[x.first] = *json_printed;